contrib/99designs/gqlgen/tracer.go

```diff
@@ -0,0 +1,96 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+// Package gqlgen contains an implementation of a gqlgen tracer, and functions to construct and configure the tracer.
+// The tracer can be passed to the gqlgen handler (see package github.com/99designs/gqlgen/handler)
+package gqlgen
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"github.com/99designs/gqlgen/graphql"
+)
+
+type trace struct {
+	graphql.NopTracer
+	serviceName   string
+	analyticsRate float64
+}
+
+// New creates an a graphql.Tracer instance that can be passed to a gqlgen handler.
+// Options can be passed in for further configuration.
+func New(opts ...Option) graphql.Tracer {
+	t := &trace{}
+	defaults(t)
+	for _, o := range opts {
+		o(t)
+	}
+	return t
+}
+
+func (t *trace) StartOperationExecution(ctx context.Context) context.Context {
+	rctx := graphql.GetRequestContext(ctx)
+	opName := ""
+	if rctx != nil && rctx.OperationName != "" {
+		opName = rctx.OperationName
+	}
+	opts := []ddtrace.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeGQLGen),
+		tracer.ResourceName(opName),
+		tracer.ServiceName(t.serviceName),
+	}
+	if !math.IsNaN(t.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, t.analyticsRate))
+	}
+	if s, ok := tracer.SpanFromContext(ctx); ok {
+		opts = append(opts, tracer.ChildOf(s.Context()))
+	}
+	_, ctx = tracer.StartSpanFromContext(ctx, opName, opts...)
+	return ctx
+}
+
+func (t *trace) StartFieldExecution(ctx context.Context, field graphql.CollectedField) context.Context {
+	span, ctx := tracer.StartSpanFromContext(ctx, "Field_"+field.Name)
+	span.SetTag("field", field.Name)
+	return ctx
+}
+
+func (t *trace) StartFieldResolverExecution(ctx context.Context, rc *graphql.ResolverContext) context.Context {
+	// This will be the span created in StartFieldExecution.
+	// StartFieldResolverExecution is called only once per StartFieldExecution, so we can add context to the
+	// span rather than starting a child span.
+	span, _ := tracer.SpanFromContext(ctx)
```

Comment:
Never ignore `_` please, anywhere.

Reply:
In this case, the `_` just tells us if a span already existed or not. It doesn't return an error or tell us that `span` is nil. (https://godoc.org/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer#SpanFromContext)

If we want, we can check it to see if span is a `NopSpan` and skip setting tags if it is, but I thought keeping the flow simpler was better, since that seems to be the purpose of `NopSpan` - return a span that does nothing rather than returning `nil` and requiring checking everywhere.

Reply:
Fair enough 👍 My bad.